File: cbsd/cbslib/routes/logs.py
# CBS service library - routes - logs
# Copyright (C) 2025  Clyso GmbH
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

from typing import Annotated

import fastapi
from cbsdcore.api.responses import BaseErrorModel, BuildLogsFollowResponse
from cbsdcore.builds.types import BuildID
from fastapi import APIRouter, Depends, HTTPException, status

from cbslib.builds.logs import BuildLogsHandlerError
from cbslib.core import utils
from cbslib.core.permissions import RoutesCaps
from cbslib.routes import logger as parent_logger
from cbslib.routes._utils import (
    CBSBuildsMgr,
    RequiredRouteCaps,
    responses_auth,
    responses_caps,
)

logger = parent_logger.getChild("logs")

router = APIRouter(prefix="/logs")


@router.get(
    "/{build_id}/tail",
    summary="Obtains the tail of the log for a given build",
    responses={
        **responses_auth,
        **responses_caps,
        500: {
            "model": BaseErrorModel,
            "description": "An internal error occurred, please check CBS service logs",
        },
        404: {
            "model": BaseErrorModel,
            "description": "Build not found",
        },
        200: {
            "model": list[str],
            "description": "List of log messages for the given build",
        },
    },
    dependencies=[Depends(RequiredRouteCaps(RoutesCaps.ROUTES_BUILDS_STATUS))],
)
async def get_build_logs_tail(
    mgr: CBSBuildsMgr,
    build_id: Annotated[BuildID, fastapi.Path(description="Build's ID")],
    n: Annotated[int, fastapi.Query(description="Number of lines to return")] = 30,
) -> BuildLogsFollowResponse:
    logger.debug(f"tail build log '{build_id}' max '{n}'")
    try:
        end_of_stream, last_id, res = await mgr.logs.tail(build_id, n)
    except BuildLogsHandlerError as e:
        logger.error(f"error handling follow request: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="check service logs for failure",
        ) from e
    except utils.FileNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="build not found",
        ) from None
    return BuildLogsFollowResponse(
        last_id=last_id, msgs=res, end_of_stream=end_of_stream
    )


@router.get(
    "/{build_id}/follow",
    summary="Follows the tail of a given build's log",
    responses={
        **responses_auth,
        **responses_caps,
        500: {
            "model": BaseErrorModel,
            "description": "An internal error occurred, please check CBS service logs",
        },
        404: {
            "model": BaseErrorModel,
            "description": "Build not found",
        },
        200: {
            "model": list[str],
            "description": "List of log messages for the given build",
        },
    },
    dependencies=[Depends(RequiredRouteCaps(RoutesCaps.ROUTES_BUILDS_STATUS))],
)
async def get_build_logs_follow(
    mgr: CBSBuildsMgr,
    build_id: Annotated[BuildID, fastapi.Path(description="Build's ID")],
    n: Annotated[int, fastapi.Query(description="Number of lines to return")] = 30,
    since: Annotated[
        str | None, fastapi.Query(description="Last message ID seen")
    ] = None,
) -> BuildLogsFollowResponse:
    logger.debug(f"follow build log '{build_id}, since '{since}', max '{n}'")
    try:
        end_of_stream, last_id, res = await mgr.logs.follow(build_id, since, max_msgs=n)
    except BuildLogsHandlerError as e:
        logger.error(f"error handling follow request: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="check service logs for failure",
        ) from e
    except utils.FileNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="build not found",
        ) from None

    return BuildLogsFollowResponse(
        last_id=last_id, msgs=res, end_of_stream=end_of_stream
    )

File: cbsd-rs/cbc/src/logs.rs
// Copyright (C) 2026  Clyso
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

//! Build log commands: tail, follow (SSE), and download.

use clap::{Args, Subcommand};
use futures_util::StreamExt;
use reqwest::Method;
use reqwest_eventsource::{Event, EventSource};
use serde::Deserialize;
use tokio::io::AsyncWriteExt;

use crate::client::CbcClient;
use crate::config::Config;
use crate::error::Error;

// ---------------------------------------------------------------------------
// CLI argument types
// ---------------------------------------------------------------------------

#[derive(Args)]
pub struct LogsArgs {
    #[command(subcommand)]
    command: LogsCommands,
}

#[derive(Subcommand)]
enum LogsCommands {
    /// Show the last N lines of a build log
    Tail(TailArgs),
    /// Follow build log output in real-time (SSE)
    Follow(FollowArgs),
    /// Download the full build log to a file
    Get(GetArgs),
}

#[derive(Args)]
struct TailArgs {
    /// Build ID
    id: i64,

    /// Number of lines to show
    #[arg(short = 'n', long = "lines", default_value = "30")]
    n: u32,
}

#[derive(Args)]
struct FollowArgs {
    /// Build ID
    id: i64,
}

#[derive(Args)]
struct GetArgs {
    /// Build ID
    id: i64,

    /// Output file path (default: build-{id}.log)
    #[arg(short, long)]
    output: Option<String>,
}

// ---------------------------------------------------------------------------
// API response types
// ---------------------------------------------------------------------------

#[derive(Deserialize)]
struct TailResponse {
    #[allow(dead_code)]
    build_id: i64,
    lines: Vec<String>,
    total_lines: u64,
    returned: u64,
}

#[derive(Deserialize)]
struct BuildStateResponse {
    #[allow(dead_code)]
    id: i64,
    state: String,
}

// ---------------------------------------------------------------------------
// Dispatch
// ---------------------------------------------------------------------------

pub async fn run(
    args: LogsArgs,
    config_path: Option<&std::path::Path>,
    debug: bool,
) -> Result<(), Error> {
    match args.command {
        LogsCommands::Tail(a) => cmd_tail(a, config_path, debug).await,
        LogsCommands::Follow(a) => cmd_follow(a, config_path, debug).await,
        LogsCommands::Get(a) => cmd_get(a, config_path, debug).await,
    }
}

// ---------------------------------------------------------------------------
// build logs tail
// ---------------------------------------------------------------------------

async fn cmd_tail(
    args: TailArgs,
    config_path: Option<&std::path::Path>,
    debug: bool,
) -> Result<(), Error> {
    let config = Config::load(config_path)?;
    let client = CbcClient::new(&config.host, &config.token, debug)?;

    let resp: TailResponse = client
        .get(&format!("builds/{}/logs/tail?n={}", args.id, args.n))
        .await?;

    println!("(showing {} of {} lines)", resp.returned, resp.total_lines);
    for line in &resp.lines {
        println!("{line}");
    }

    Ok(())
}

// ---------------------------------------------------------------------------
// build logs follow
// ---------------------------------------------------------------------------

/// Maximum consecutive SSE errors before giving up.
const MAX_RETRIES: u32 = 3;

async fn cmd_follow(
    args: FollowArgs,
    config_path: Option<&std::path::Path>,
    debug: bool,
) -> Result<(), Error> {
    let config = Config::load(config_path)?;
    let client = CbcClient::new(&config.host, &config.token, debug)?;

    let request =
        client.request_builder(Method::GET, &format!("builds/{}/logs/follow", args.id))?;

    let mut es = EventSource::new(request)
        .map_err(|e| Error::Other(format!("cannot create event source: {e}")))?;

    let mut retries: u32 = 0;

    while let Some(event) = es.next().await {
        match event {
            Ok(Event::Message(msg)) if msg.event == "done" => {
                break;
            }
            Ok(Event::Message(msg)) => {
                println!("{}", msg.data);
                retries = 0;
            }
            Ok(Event::Open) => {
                retries = 0;
            }
            Err(e) => {
                retries += 1;
                if retries >= MAX_RETRIES {
                    eprintln!("stream error: {e}");
                    es.close();
                    break;
                }
            }
        }
    }

    // Fetch the final build state.
    let build: BuildStateResponse = client.get(&format!("builds/{}", args.id)).await?;

    println!("--- build {} finished: {} ---", args.id, build.state);

    Ok(())
}

// ---------------------------------------------------------------------------
// build logs get
// ---------------------------------------------------------------------------

async fn cmd_get(
    args: GetArgs,
    config_path: Option<&std::path::Path>,
    debug: bool,
) -> Result<(), Error> {
    let config = Config::load(config_path)?;
    let client = CbcClient::new(&config.host, &config.token, debug)?;

    let output_path = args
        .output
        .unwrap_or_else(|| format!("build-{}.log", args.id));

    let mut resp = client
        .get_stream(&format!("builds/{}/logs", args.id))
        .await?;

    let mut file = tokio::fs::File::create(&output_path)
        .await
        .map_err(|e| Error::Other(format!("cannot create '{output_path}': {e}")))?;

    let mut total_bytes: u64 = 0;

    while let Some(chunk) = resp
        .chunk()
        .await
        .map_err(|e| Error::Connection(format!("download error: {e}")))?
    {
        file.write_all(&chunk)
            .await
            .map_err(|e| Error::Other(format!("write error: {e}")))?;
        total_bytes += chunk.len() as u64;
    }

    println!(
        "log written to {output_path} ({})",
        format_size(total_bytes)
    );

    Ok(())
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/// Format a byte count as a human-readable string.
fn format_size(bytes: u64) -> String {
    const KB: u64 = 1024;
    const MB: u64 = 1024 * KB;
    const GB: u64 = 1024 * MB;

    if bytes >= GB {
        format!("{:.1} GB", bytes as f64 / GB as f64)
    } else if bytes >= MB {
        format!("{:.1} MB", bytes as f64 / MB as f64)
    } else if bytes >= KB {
        format!("{:.1} KB", bytes as f64 / KB as f64)
    } else {
        format!("{bytes} B")
    }
}
